### PR TITLE
Canonicalize relationals as well

### PR DESCRIPTION
**Pull request by MaxGraey**

### CODE REVIEW

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 20131
   CompareVMs: 3557
   CheckDeterminism: 1754
   Wasm2JS: 2095
   Asyncify: 3186

ITERATION: 22163
```

**Comment by MaxGraey** [comment]
@kripken @tlively could you take a look at this?

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -895,9 +895,13 @@ struct OptimizeInstructions
   // Canonicalizing the order of a symmetric binary helps us
   // write more concise pattern matching code elsewhere.
   void canonicalize(Binary* binary) {
-    assert(isSymmetric(binary));
+    assert(isSymmetricOrRelational(binary));
+    bool isRelational = binary->isRelational();
     auto swap = [&]() {
       assert(canReorder(binary->left, binary->right));
+      if (isRelational) {
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:902]:
```suggestion
      if (binary->isRelational()) {
```
This will let you avoid needing a new variable.

**Review by tlively — APPROVED** [review]

**Review by tlively — APPROVED** [review]

**Comment by kripken** [comment]
Looks like this caused a bug, the fuzzer found `630869400013041980`,
```wat
(module
 (export "func_150" (func $0))
 (func $0 (param $0 i64) (result i32)
  (i64.ge_s
   (i64.const 0)
   (i64.sub
    (i64.const 4294967263)
    (block $label$1 (result i64)
     (i64.const -9223372036854775808)
    )
   )
  )
 )
)
```
STR: `wasm-opt --optimize-instructions --fuzz-exec`

**Comment by MaxGraey** [comment]
> Looks like this caused a bug, the fuzzer found 630869400013041980,

Hmm, that's condition should return `true` (1) and it seems this right. Also I can't reproduce this on master branch with `630869400013041980` seed. Are you sure it's on the master?

**Comment by kripken** [comment]
I am on a side branch - sorry, I didn't think it added anything that could matter here.

But the problem happens with the steps to reproduce from above (`wasm-opt --optimize-instructions --fuzz-exec` on that wat), and it bisects to this PR.

**Comment by MaxGraey** [comment]
Oh, that's happening only on side branch? Hmm, not sure I understand what exactly going wrong in this case. If you have see some problem with changes introduced in this PR let me know and I try to fix this.

**Comment by kripken** [comment]
Sorry if I wasn't clear. This happens on master. It's just the fuzz seed that was from a side branch.

I think I found the bug meanwhile:
```wat
;; ok
(i64.le_s
 (i64.sub
  (i64.const 4294967263)
  (block $label$1 (result i64)
   (i64.const -9223372036854775808)           
  )
 )
 (i64.const 0)
)

=>

;; bad
(i64.le_s
 (i64.const 4294967263)
 (block $label$1 (result i64)
  (i64.const -9223372036854775808)           
 )
)
```

It is basically doing this:
```
X - Y <= 0

=>

X <= Y
```

That is true mathematically, but not in the case of an overflow, e.g. `X=10, Y=0x8000000000000000`. `X - Y` is a negative number, so `X - Y <= 0` is true. But it is not true that `X <= Y` (as Y is negative, but X is not).

 . So this code looks wrong:
```cpp
        // signed(x - y) <=> 0  =>  x <=> y
        if (matches(curr,
                    binary(&op,
                           binary(&inner, Abstract::Sub, any(), any()),
                           ival(0))) &&
            (op == Abstract::getBinary(type, Abstract::Eq) ||
             op == Abstract::getBinary(type, Abstract::Ne) ||
             op == Abstract::getBinary(type, Abstract::LeS) ||
             op == Abstract::getBinary(type, Abstract::LtS) ||
             op == Abstract::getBinary(type, Abstract::GeS) ||
             op == Abstract::getBinary(type, Abstract::GtS))) {
          curr->right = inner->right;
          curr->left = inner->left;
          return curr;
        }
```
(from #3275) - this PR just uncovered the bug, it seems, but flipping the order.

**Comment by MaxGraey** [comment]
Hmm, now I see. It seesm we need special handling for `x - 0x8000000000000000 < 0` and `x - 0x8000000000000000 <= 0`

**Comment by MaxGraey** [comment]
Hmm, `i32(x) - i32(0x80000000) < 0` should already skipped due canonicalization to `i32(x) + i32(0x80000000) < 0`. So I'm still wondering why this failed.

**Comment by kripken** [comment]
Note the `block` there: it prevents it from changing `X - C => X + (-C)` because there isn't a const there, there is a block.

**Comment by MaxGraey** [comment]
> Note the block there: it prevents it from changing X - C => X + (-C)

Good point!

**Comment by MaxGraey** [comment]
Also I'm wondering for practical this? Due to fuzzer use pretty syntactic case when `--optimize-instructions`  used in solo without pre-simplified CFG which remove block with single constant member

**Comment by kripken** [comment]
It's true that a block like that might be removed. But here is another variation, with a global:
```wat
(module
 (global $g i32 (i32.const 0x80000000))
 (func "foo" (param $0 i32) (result i32)
  (i32.ge_s
   (i32.const 0)
   (i32.sub
    (i32.const 8)
    (global.get $g)
   )
  )
 )
)
```
The block just prevents canonicalization and other things. But the value coming from a global, or a call, etc., would do similar things - in all those cases we'd only see the value at runtime. So we have to fix this

**Comment by MaxGraey** [comment]
It seems we shoudn't apply this transform for general case like this done in LLVM. For example GCC don't apply it:
https://godbolt.org/z/7Gf885

**Comment by kripken** [comment]
I don't know enough about x86 to know if clang is right there or not. But changing to wasm output on compiler explorer, it doesn't do the optimization, confirming what we see here.

Let's remove it. I can make a PR unless you are already working on one @MaxGraey ?

**Comment by MaxGraey** [comment]
I already done this on #3336

**Comment by kripken** [comment]
Oh wait, I was reading that wrong. It does do it, hmm...

**Comment by MaxGraey** [comment]
Yes, I guess LLVM can't do this. Usually GCC more accurate with such things.

**Pull request merged**